cmd/dp-hierarchy-builder/main.go

```diff
@@ -2,127 +2,194 @@ package main
+	if err != nil {
+		log.Event(ctx, "error creating kafka producer", log.FATAL, log.Error(err))
+		os.Exit(1)
+	}
 
-	kafkaErrorProducer, err := kafka.NewProducer(cfg.KafkaAddr, cfg.ErrorProducerTopic, 0)
-	exitIfError(err)
+	errorProducerChannels := kafka.CreateProducerChannels()
+	kafkaErrorProducer, err := kafka.NewProducer(ctx, kafkaBrokers, cfg.ErrorProducerTopic, useDefaultMaxMessageSize, errorProducerChannels)
+	exitIfError(ctx, err, "error creating kafka producer")
 
 	avroProducer := event.NewAvroProducer(kafkaProducer)
 
-	db, err := graph.NewHierarchyStore(context.Background())
-	exitIfError(err)
+	db, err := graph.NewHierarchyStore(ctx)
+	exitIfError(ctx, err, "error creating Kafka error producer")
 
 	// when errors occur - we send a message on an error topic.
 	errorHandler, err := reporter.NewImportErrorReporter(kafkaErrorProducer, log.Namespace)
-	exitIfError(err)
+	exitIfError(ctx, err, "error creating import error reporter")
 
 	eventHandler := event.NewDataImportCompleteHandler(&hierarchy.Store{db}, avroProducer)
 
 	eventConsumer := event.NewConsumer()
-	eventConsumer.Consume(kafkaConsumer, eventHandler, errorHandler)
-
-	healthChecker := healthcheck.NewServer(
-		cfg.BindAddr,
-		cfg.HealthCheckInterval,
-		cfg.HealthCheckRecoveryInterval,
-		errorChannel,
-		db,
-	)
+	eventConsumer.Consume(ctx, kafkaConsumer, eventHandler, errorHandler)
+
+	hc := startHealthChecks(ctx, cfg, kafkaConsumer, kafkaProducer, kafkaErrorProducer, db)
+
+	apiErrors, httpServer := startApi(ctx, hc, cfg)
 
 	signals := make(chan os.Signal, 1)
 	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
 
 	// this will block (main) until a fatal error occurs
 	select {
-	case err := <-kafkaConsumer.Errors():
-		log.ErrorC("kafka consumer", err, nil)
-	case err := <-kafkaProducer.Errors():
-		log.ErrorC("kafka result producer", err, nil)
-	case err := <-kafkaErrorProducer.Errors():
-		log.ErrorC("kafka error producer", err, nil)
+	case err := <-apiErrors:
+		log.Event(ctx, "http server error", log.ERROR, log.Error(err))
+	case err := <-kafkaConsumer.Channels().Errors:
+		log.Event(ctx, "kafka consumer error", log.ERROR, log.Error(err))
+	case err := <-kafkaProducer.Channels().Errors:
+		log.Event(ctx, "kafka result producer error", log.ERROR, log.Error(err))
+	case err := <-kafkaErrorProducer.Channels().Errors:
+		log.Event(ctx, "kafka error producer error", log.ERROR, log.Error(err))
 	case err := <-errorChannel:
-		log.ErrorC("error channel", err, nil)
+		log.Event(ctx, "error channel error", log.ERROR, log.Error(err))
 	case <-signals:
-		log.Debug("os signal received", nil)
+		log.Event(ctx, "os signal received", log.INFO)
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+	ctx, cancel := context.WithTimeout(ctx, cfg.GracefulShutdownTimeout)
 
 	// gracefully dispose resources
+	hc.Stop()
+
+	err = httpServer.Close(ctx)
+	logIfError(ctx, err, "error closing http server")
+
 	err = eventConsumer.Close(ctx)
```

Comment:
If the event consumer was not initialised on start up this will panic or throw an error that it couldn't close something that doesn't exist.

This is why some of the applications David and I have made changes to have an initialise package for these dependencies to track if they have successfully initialised so we know to close them later. See observation-importer

Reply:
Wouldn't you expect an error to occur if it wasn't initialised on startup, causing it to exit?

Reply:
Only if there was a configuration error, missing necessary configuration like brokers etc. But following our conversation I think you are right that it should be up to the kafka library to handle this not the application.